// @1inch/cross-chain-sdk/dist/cjs/api/quoter/quote/quote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Quote = void 0;
const tslib_1 = require("tslib");
const byte_utils_1 = require("@1inch/byte-utils");
const fusion_sdk_1 = require("@1inch/fusion-sdk");
const assert_1 = tslib_1.__importDefault(require("assert"));
const addresses_1 = require("../../../domains/addresses");
const time_locks_1 = require("../../../domains/time-locks");
const types_1 = require("../types");
const preset_1 = require("../preset");
const evm_1 = require("../../../cross-chain-order/evm");
const chains_1 = require("../../../chains");
class Quote {
    // eslint-disable-next-line max-params
    constructor(params, quoteId, srcTokenAmount, dstTokenAmount, presets, srcEscrowFactory, dstEscrowFactory, timeLocks, srcSafetyDeposit, dstSafetyDeposit, whitelist, recommendedPreset, prices, volume, slippage) {
        this.params = params;
        this.quoteId = quoteId;
        this.srcTokenAmount = srcTokenAmount;
        this.dstTokenAmount = dstTokenAmount;
        this.presets = presets;
        this.srcEscrowFactory = srcEscrowFactory;
        this.dstEscrowFactory = dstEscrowFactory;
        this.timeLocks = timeLocks;
        this.srcSafetyDeposit = srcSafetyDeposit;
        this.dstSafetyDeposit = dstSafetyDeposit;
        this.whitelist = whitelist;
        this.recommendedPreset = recommendedPreset;
        this.prices = prices;
        this.volume = volume;
        this.slippage = slippage;
    }
    get srcChainId() {
        return this.params.srcChain;
    }
    get dstChainId() {
        return this.params.dstChain;
    }
    static fromEVMQuote(request, response) {
        const presets = {
            [types_1.PresetEnum.fast]: new preset_1.Preset(response.presets.fast),
            [types_1.PresetEnum.medium]: new preset_1.Preset(response.presets.medium),
            [types_1.PresetEnum.slow]: new preset_1.Preset(response.presets.slow),
            [types_1.PresetEnum.custom]: response.presets.custom
                ? new preset_1.Preset(response.presets.custom)
                : undefined
        };
        const dstEscrowFactory = (0, chains_1.isEvm)(request.dstChain)
            ? addresses_1.EvmAddress.fromString(response.dstEscrowFactory)
            : addresses_1.SolanaAddress.fromString(response.dstEscrowFactory);
        return new Quote(request, response.quoteId, BigInt(response.srcTokenAmount), BigInt(response.dstTokenAmount), presets, addresses_1.EvmAddress.fromString(response.srcEscrowFactory), dstEscrowFactory, response.timeLocks, BigInt(response.srcSafetyDeposit), BigInt(response.dstSafetyDeposit), response.whitelist.map((w) => addresses_1.EvmAddress.fromString(w)), response.recommendedPreset, response.prices, response.volume, response.autoK);
    }
    static fromSolanaQuote(request, response) {
        const presets = {
            [types_1.PresetEnum.fast]: new preset_1.Preset(response.presets.fast),
            [types_1.PresetEnum.medium]: new preset_1.Preset(response.presets.medium),
            [types_1.PresetEnum.slow]: new preset_1.Preset(response.presets.slow),
            [types_1.PresetEnum.custom]: response.presets.custom
                ? new preset_1.Preset(response.presets.custom)
                : undefined
        };
        const dstEscrowFactory = (0, chains_1.isEvm)(request.dstChain)
            ? addresses_1.EvmAddress.fromString(response.dstEscrowFactory)
            : addresses_1.SolanaAddress.fromString(response.dstEscrowFactory);
        return new Quote(request, response.quoteId, BigInt(response.srcTokenAmount), BigInt(response.dstTokenAmount), presets, addresses_1.SolanaAddress.fromString(response.srcEscrowFactory), dstEscrowFactory, response.timeLocks, BigInt(response.srcSafetyDeposit), BigInt(response.dstSafetyDeposit), [], response.recommendedPreset, response.prices, response.volume, response.autoK);
    }
    createEvmOrder(params) {
        (0, assert_1.default)(this.isEvmQuote(), 'cannot create non evm order');
        const preset = this.getPreset(params?.preset || this.recommendedPreset);
        const auctionDetails = preset.createAuctionDetails(params.delayAuctionStartTimeBy);
        const allowPartialFills = preset.allowPartialFills;
        const allowMultipleFills = preset.allowMultipleFills;
        const isNonceRequired = !allowPartialFills || !allowMultipleFills;
        const nonce = isNonceRequired
            ? (params.nonce ?? (0, fusion_sdk_1.randBigInt)(byte_utils_1.UINT_40_MAX))
            : params.nonce;
        const takerAsset = this.params.dstTokenAddress.isNative()
            ? addresses_1.EvmAddress.NATIVE
            : this.params.dstTokenAddress;
        return evm_1.EvmCrossChainOrder.new(this.srcEscrowFactory, {
            makerAsset: this.params.srcTokenAddress,
            takerAsset: takerAsset,
            makingAmount: this.srcTokenAmount,
            takingAmount: preset.auctionEndAmount,
            maker: this.params.walletAddress,
            receiver: params.receiver
        }, {
            hashLock: params.hashLock,
            srcChainId: this.params.srcChain,
            dstChainId: this.params.dstChain,
            srcSafetyDeposit: this.srcSafetyDeposit,
            dstSafetyDeposit: this.dstSafetyDeposit,
            timeLocks: time_locks_1.TimeLocks.new({
                srcWithdrawal: BigInt(this.timeLocks.srcWithdrawal),
                srcPublicWithdrawal: BigInt(this.timeLocks.srcPublicWithdrawal),
                srcCancellation: BigInt(this.timeLocks.srcCancellation),
                srcPublicCancellation: BigInt(this.timeLocks.srcPublicCancellation),
                dstWithdrawal: BigInt(this.timeLocks.dstWithdrawal),
                dstPublicWithdrawal: BigInt(this.timeLocks.dstPublicWithdrawal),
                dstCancellation: BigInt(this.timeLocks.dstCancellation)
            })
        }, {
            auction: auctionDetails,
            whitelist: this.getWhitelist(auctionDetails.startTime, preset.exclusiveResolver)
        }, {
            nonce,
            permit: params.permit,
            allowPartialFills,
            allowMultipleFills,
            orderExpirationDelay: params?.orderExpirationDelay,
            source: this.params.source,
            enablePermit2: params.isPermit2
        });
    }
    // todo: create Solana order
    isEvmQuote() {
        return (0, chains_1.isEvm)(this.params.srcChain);
    }
    isSolanaQuote() {
        return (0, chains_1.isSolana)(this.params.srcChain);
    }
    getPreset(type = this.recommendedPreset) {
        return this.presets[type];
    }
    getWhitelist(auctionStartTime, exclusiveResolver) {
        if (exclusiveResolver) {
            return this.whitelist.map((resolver) => {
                const isExclusive = resolver.equal(exclusiveResolver);
                return {
                    address: resolver,
                    allowFrom: isExclusive ? 0n : auctionStartTime
                };
            });
        }
        return this.whitelist.map((resolver) => ({
            address: resolver,
            allowFrom: 0n
        }));
    }
}
exports.Quote = Quote;
